packages/react/src/components/login/beacon/signals.ts

```diff
@@ -0,0 +1,370 @@
+    },
+    namespace,
+  };
+}
+
+// payload and context are conditionally added, so they can be used as triggers
+// if triggers have object[key], the value is checked
+export function createSignalOrTriggerProps(
+  type: SignalType,
+  namespace: SignalNamespace,
+  payload?: SignalPayload,
+  context?: SignalContext,
+): Signal {
+  return {
+    type,
+    namespace,
+    ...(payload && { payload }),
+    ...(context && { context }),
+  };
+}
+
+export function createErrorSignal(namespace: SignalNamespace, payload?: ErrorPayload): ErrorSignal {
+  return createSignalOrTriggerProps(errorSignalType, namespace, payload) as ErrorSignal;
+}
+
+export function createEventSignal(namespace: SignalNamespace, payload?: EventPayload): EventSignal {
+  return createSignalOrTriggerProps(eventSignalType, namespace, payload) as EventSignal;
+}
+
+export function createStateChangeSignal(
+  namespace: SignalNamespace,
+  payload?: StateChangeSignalPayload,
+): StateChangeSignal {
+  return createSignalOrTriggerProps(stateChangeSignalType, namespace, payload) as Signal<
+    typeof stateChangeSignalType,
+    StateChangeSignalPayload
+  >;
+}
+
+export function createErrorTriggerProps(
+  namespace: SignalNamespace = LISTEN_TO_ALL_MARKER,
+  type?: string,
+): Pick<Signal, 'namespace'> & { type: ErrorSignal['type'] } & { payload?: { type: string } } {
+  return {
+    type: errorSignalType,
+    namespace,
+    ...(type && { payload: { type } }),
+  };
+}
+
+export function createEventTriggerProps(
+  namespace: SignalNamespace = LISTEN_TO_ALL_MARKER,
+  type?: string,
+): Pick<Signal, 'namespace'> & { type: EventSignal['type'] } & { payload?: { type: string } } {
+  return {
+    type: eventSignalType,
+    namespace,
+    ...(type && { payload: { type } }),
+  };
+}
+
+export function createInitTriggerProps(
+  namespace: SignalNamespace = LISTEN_TO_ALL_MARKER,
+): Pick<Signal, 'type' | 'namespace'> {
+  return {
+    type: initSignalType,
+    namespace,
+  };
+}
+
+export function createStateChangeTriggerProps(
+  namespace: SignalNamespace = LISTEN_TO_ALL_MARKER,
+  state?: string,
+  previousState?: string,
+): Pick<Signal, 'type' | 'namespace'> {
+  const payload = state
+    ? {
+        ...(state && { state }),
+        ...(previousState && { previousState }),
+      }
+    : undefined;
+  return {
+    type: stateChangeSignalType,
+    namespace,
+    ...(payload && { payload }),
+  };
+}
+
+export function createTriggerPropsForAllSignals(
+  namespace: SignalNamespace = LISTEN_TO_ALL_MARKER,
+): Pick<Signal, 'type' | 'namespace'> {
+  return {
+    type: LISTEN_TO_ALL_MARKER,
+    namespace,
+  };
+}
+
+export function createTriggerForAllSignalTypes(): Pick<Signal, 'type'> {
+  return {
+    type: LISTEN_TO_ALL_MARKER,
+  };
+}
+
+export function createTriggerForAllNamespaces(): Pick<Signal, 'namespace'> {
+  return {
+    namespace: LISTEN_TO_ALL_MARKER,
+  };
+}
```

Comment:
Looks compact, but harder to read IMHO
```
    ...(payloadType && { payload: { type: payloadType } }),
    ...(payload && { payload }),
```
Also the 5 args, should probably be passed as one object. If I remember correctly, I tried to avoid one generic function, because then there are so many unused arguments (or object keys) for each scenario.